terraform/context_apply_test.go

```diff
@@ -2639,105 +2639,107 @@ module.child:
   `)
 }
 
-func TestContext2Apply_moduleOrphanProvider(t *testing.T) {
-	m := testModule(t, "apply-module-orphan-provider-inherit")
-	p := testProvider("aws")
-	p.ApplyFn = testApplyFn
-	p.DiffFn = testDiffFn
-
-	p.ConfigureFn = func(c *ResourceConfig) error {
-		if _, ok := c.Get("value"); !ok {
-			return fmt.Errorf("value is not found")
-		}
-
-		return nil
-	}
-
-	// Create a state with an orphan module
-	state := &State{
-		Modules: []*ModuleState{
-			&ModuleState{
-				Path: []string{"root", "child"},
-				Resources: map[string]*ResourceState{
-					"aws_instance.bar": &ResourceState{
-						Type: "aws_instance",
-						Primary: &InstanceState{
-							ID: "bar",
-						},
-					},
-				},
-			},
-		},
-	}
-
-	ctx := testContext2(t, &ContextOpts{
-		Module: m,
-		State:  state,
-		ProviderResolver: ResourceProviderResolverFixed(
-			map[string]ResourceProviderFactory{
-				"aws": testProviderFuncFixed(p),
-			},
-		),
-	})
-
-	if _, err := ctx.Plan(); err != nil {
-		t.Fatalf("err: %s", err)
-	}
-
-	if _, err := ctx.Apply(); err != nil {
-		t.Fatalf("err: %s", err)
-	}
-}
-
-func TestContext2Apply_moduleOrphanGrandchildProvider(t *testing.T) {
-	m := testModule(t, "apply-module-orphan-provider-inherit")
-	p := testProvider("aws")
-	p.ApplyFn = testApplyFn
-	p.DiffFn = testDiffFn
-
-	p.ConfigureFn = func(c *ResourceConfig) error {
-		if _, ok := c.Get("value"); !ok {
-			return fmt.Errorf("value is not found")
-		}
-
-		return nil
-	}
-
-	// Create a state with an orphan module that is nested (grandchild)
-	state := &State{
-		Modules: []*ModuleState{
-			&ModuleState{
-				Path: []string{"root", "parent", "child"},
-				Resources: map[string]*ResourceState{
-					"aws_instance.bar": &ResourceState{
-						Type: "aws_instance",
-						Primary: &InstanceState{
-							ID: "bar",
-						},
-					},
-				},
-			},
-		},
-	}
-
-	ctx := testContext2(t, &ContextOpts{
-		Module: m,
-		State:  state,
-		ProviderResolver: ResourceProviderResolverFixed(
-			map[string]ResourceProviderFactory{
-				"aws": testProviderFuncFixed(p),
-			},
-		),
-	})
-
-	if _, err := ctx.Plan(); err != nil {
-		t.Fatalf("err: %s", err)
-	}
-
-	if _, err := ctx.Apply(); err != nil {
-		t.Fatalf("err: %s", err)
-	}
-}
+//// FIXME: how do we handle this one?
```

Comment:
I think this example, as given, should ideally be an explicit plan-time failure that the provider that created this module is no longer available, and that the user must put it back to complete the destroy.

A different example where the module is orphaned but its provider was originally defined in the root would then work as expected.

However, since this never worked properly before I think it's fine to save it for a subsequent change, as long as the design here isn't painting us into a corner that would make it hard to reach this end-state. (I don't _think_ it is... :thinking: )

Reply:
The problem here is that the only way to remove modules (as of this PR) is to recreate the module tree with only the providers. Removing a module with multiple levels of nested modules could be quite painful. Now that we're pushing modules for composition, I think this puts us pretty far from an acceptable solution (it at least kind-of could work before). 

We could rig something with the new module providers block, but then I'm not sure what to do with the [great]*-grandchild orphaned resources. Going through the work of having a root level provider adopt them should bring back the original behavior, but we still eventually need a way to bind them to a correctly configured provider.

More ideas I've had:
- Store the provider along with a serialized version of its config when it last modified the resource in the resource state. Big and ugly, but it might just work.
- Require the use of a targeted destroy first, or make a special command for removing a module.